etcdserver/server.go

```diff
@@ -480,21 +480,25 @@ type etcdProgress struct {
 	appliedi  uint64
 }
 
-// newApplier buffers apply operations and streams their results over an
-// etcdProgress output channel. This is so raftNode won't block on sending
+// newApplier buffers apply operations so raftNode won't block on sending
 // new applies, timing out (since applies can be slow). The goroutine begins
-// shutdown on close(s.done) and closes the etcdProgress channel when finished.
-func (s *EtcdServer) newApplier(ep etcdProgress) <-chan etcdProgress {
-	etcdprogc := make(chan etcdProgress)
+// shutdown on close(s.done) and closes the returned channel when finished.
+func (s *EtcdServer) startApplier(ep etcdProgress) <-chan struct{} {
+	donec := make(chan struct{})
 	go func() {
-		defer close(etcdprogc)
+		defer close(donec)
 		pending := []apply{}
 		sdonec := s.done
 		apdonec := make(chan struct{})
 		// serialized function
 		f := func(ap apply) {
 			s.applyAll(&ep, &ap)
-			etcdprogc <- ep
+			select {
+			case m := <-s.msgSnapC:
```

Comment:
i feel it is cleaner and more efficient (no two select for one apply) to move this to into the select loop at line 505. I think that is also safe.